### PR TITLE
[feat]curvefs: change hostname to optional

### PR DESCRIPTION
**Pull request by 201341**
### What problem does this PR solve?

Issue Number: #2493  

Problem Summary:

### What is changed and how it works?

What's Changed:

How it Works:

Side effects(Breaking backward compatibility? Performance regression?):

### Check List

- [x] Relevant documentation/comments is changed or added
- [x] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

**Comment by 201341** [comment]
cicheck

File: curvefs/src/metaserver/register.cpp [hunk 1]
```diff
@@ -62,11 +62,10 @@ int Register::RegisterToMDS(MetaServerMetadata *metadata) {
     char hostname[128];
     int ret = gethostname(hostname, sizeof(hostname));
     if (ret == -1) {
-        LOG(ERROR) << "get hostname failed!";
-        return -1;
+        LOG(INFO) << "get hostname failed!";
+    } else {
+        req.set_hostname(hostname);
```
Comment (SeanHai) [inline curvefs/src/metaserver/register.cpp:67]: You need change the hostName of MetaServerRegistRequest in topology.proto to optional.

File: curvefs/src/tools/create/curvefs_create_topology_tool.cpp [hunk 2]
```diff
@@ -628,7 +628,9 @@ int CurvefsBuildTopologyTool::CreateServer() {
     for (auto it : serverDatas) {
         ServerRegistRequest request;
         ServerRegistResponse response;
-        request.set_hostname(it.name);
+        if (!it.name.empty()) {
+            request.set_hostname(it.name);
```
Comment (SeanHai) [inline curvefs/src/tools/create/curvefs_create_topology_tool.cpp:632]: You need change hostName of ServerRegistRequest in topology.topo to optional.

File: curvefs/src/mds/topology/topology_manager.cpp [hunk 3]
```diff
@@ -125,7 +125,12 @@ void TopologyManager::RegistMetaServer(const MetaServerRegistRequest *request,
         }
     }
 
-    MetaServer metaserver(metaServerId, request->hostname(), token, serverId,
+    std::string hostname;
+    if (request->has_hostname()) {
```
Comment (SeanHai) [inline curvefs/src/mds/topology/topology_manager.cpp:129]: not need, request->hostname() is empty string if hostname not set.

**Comment by SeanHai** [comment]
You should check all code that uses hostname to see if remove hostname affects it.
For example: https://github.com/opencurve/curve/blob/ace571a80e90c12239b9a8ba0dbc0a230090bd3f/curvefs/src/mds/topology/topology.cpp#L1654

File: curvefs/proto/topology.proto [hunk 4]
```diff
@@ -79,7 +79,7 @@ message ZoneData {
 
 message ServerData {
     required uint32 serverId = 1;
-    required string hostName = 2;
+    optional string hostName = 2;
```
[the remark this answers is not in the document]
Comment (201341) [reply curvefs/proto/topology.proto:82]: Yes, I will change the curvebs code in another pr for passing ci.

File: curvefs/src/metaserver/register.cpp (hunk 1 above)
Comment (201341) [reply curvefs/src/metaserver/register.cpp:67]: ok

File: curvefs/src/tools/create/curvefs_create_topology_tool.cpp (hunk 2 above)
Comment (201341) [reply curvefs/src/tools/create/curvefs_create_topology_tool.cpp:632]: ok

File: curvefs/src/mds/topology/topology_manager.cpp (hunk 3 above)
Comment (201341) [reply curvefs/src/mds/topology/topology_manager.cpp]: ok

**Comment by 201341** [comment]
cicheck

**Comment by SeanHai** [comment]
> 1. You should check all code that uses hostname to see if remove hostname affects it.
>    For example: https://github.com/opencurve/curve/blob/ace571a80e90c12239b9a8ba0dbc0a230090bd3f/curvefs/src/mds/topology/topology.cpp#L1654
> 2. You need add some unit test of you modified when not set hostname.


 The modify of hostname is not only affect register, you need check other place used hostname  like above example.

**Comment by 201341** [comment]
> > 1. You should check all code that uses hostname to see if remove hostname affects it.
> >    For example: https://github.com/opencurve/curve/blob/ace571a80e90c12239b9a8ba0dbc0a230090bd3f/curvefs/src/mds/topology/topology.cpp#L1654
> > 2. You need add some unit test of you modified when not set hostname.
> 
> The modify of hostname is not only affect register, you need check other place used hostname like above example.

Yes, I find no other codes that effect by hostname after search keyword 'hostname'.

**Review by SeanHai — APPROVED** [review]

**Comment by SeanHai** [comment]
LGTM!

**Review by ilixiaocui — APPROVED** [review]

**Pull request merged**